index.js

```diff
@@ -106,7 +106,7 @@ module.exports = fp(function from (fastify, opts, next) {
         if (!this.sent) {
           if (err.code === 'ERR_HTTP2_STREAM_CANCEL' || err.code === 'ENOTFOUND') {
             onError(this, { error: new createError.ServiceUnavailable() })
-          } else if (err instanceof TimeoutError || err.code === 'UND_ERR_REQUEST_TIMEOUT') {
+          } else if (err instanceof TimeoutError || err.code === 'UND_ERR_HEADERS_TIMEOUT' || err.code === 'UND_ERR_BODY_TIMEOUT') {
```

Comment:
Do you think `UND_ERR_BODY_TIMEOUT` could happen here? From what I've read it will happen after we have started streaming the response.

Reply:
I think you're right, but I'm going to need to read some more of undici's source code to understand the flow here. I'll pick this up tomorrow morning.

Reply:
i.e.

https://github.com/nodejs/undici/blob/c4e237067dbc784d1cefa2b80a3ca5213a3561c8/lib/core/client.js#L731-L735

https://github.com/nodejs/undici/blob/c4e237067dbc784d1cefa2b80a3ca5213a3561c8/lib/core/client.js#L522-L524      ? setTimeout(onBodyTimeout, client[kBodyTimeout], this)
      : null

Reply:
I think more thought is needed here.

If a user sets `bodyTimeout` but not `headersTimeout`, what would they expect the behaviour to be? And what are our options on handling this?

There are a few edge cases that would need to be considered and perhaps tested for:


## Case 1

The headers take longer to respond than `bodyTimeout`, but the subsequent time taken to send the body being less than `bodyTimeout`.

This should complete successfully, the headers can take as long as undici allows them to, `bodyTimeout` is the time taken from the headers being sent to the body being closed (`res.end()`).

I can add a test for this.

## Case 2

The headers are sent quickly, but the body is not sent at all within the `bodyTimeout`

At the moment, we send response code 500 to the client, along with the error from Undici in the body.

https://github.com/penx/fastify-reply-from/blob/bf4793159d1864f36e366c366098e60eccee99bd/index.js#L130-L130

`stream.destroy(<BodyTimeoutError>)`
https://github.com/nodejs/undici/blob/7f3eef73355ced497107023a7af8b987fd01459c/lib/core/util.js#L105

```json
{
  statusCode: 500,
  code: 'UND_ERR_BODY_TIMEOUT',
  error: 'Internal Server Error',
  message: 'Body Timeout Error'
}
```

## Case 3

The headers being sent quickly, but the body being *partially sent* within the `bodyTimeout` period.

```js
const target = http.createServer((req, res) => {
  t.pass('request proxied')
  req.on('data', () => undefined)
  req.on('end', () => {
    res.flushHeaders()
    res.write('test')
    clock.setTimeout(() => {
      res.end()
      t.end()
    }, 1)
  })
})
```

This partially streams a response ("test") to the client and then hits node internals `stream.destroy(<BodyTimeoutError>)` (as in Case 2): 

https://github.com/nodejs/undici/blob/7f3eef73355ced497107023a7af8b987fd01459c/lib/core/util.js#L105

However on this occasion, node internals seem to behave differently because the body has already been partially streamed.

`got.get` throws a "The server aborted pending request" error, and reports that the response code is '200'. `got` does not allow us to access the partially streamed response, as far as I can see.

Confirmed looking at a local proxy that the response code remains as 200 and "test" is sent to the client.

---

I don't think the behaviour of 2 and 3 are great, but I'm going to need to do some more reading to figure out if we have any options to improve them.

Reply:
> Do you think UND_ERR_BODY_TIMEOUT could happen here? From what I've read it will happen after we have started streaming the response.

Perhaps if the response code changes to 500 before we start streaming the body to the client, we could intercept it 🤷 

But as the code stands at the moment, no I don't think `UND_ERR_BODY_TIMEOUT` would be intercepted.

Reply:
As for the HTTP spec, headers are alway sent _before_ the body. If they are _after_ the body are called trailers. I don't think this module supports trailers.

Reply:
does that include any response code change that is sent after `flushHeaders` but before anything is written to the body?

Reply:
ah I've been using `res.flushHeaders()` on the fake servers when I should probably be using `res.writeHead(200)` or similar, which changes things a bit

Reply:
ok so in case 2, the headers have not yet been sent to the client, so when an error is received from the target server, and we call `payload.pipe(res)`, something in node internals seems to change the statusCode from 200 to 500.

In case 3, the headers have already been sent to the client, and so node has no choice but to terminate the response.

We may be able to do something in case 2 but I'd need to read the node internals code to see what it's doing when it changes the status to 500.

Reply:
Read up on https://developer.mozilla.org/en-US/docs/Web/HTTP/Messages.

TL;DR the order is start lines (status code, etc), headers, body, trailers.

> In case 3, the headers have already been sent to the client, and so node has no choice but to terminate the response (well I guess it could send a trailer as you say, but it doesn't seem to).

Yes. The response has already been sent. You'll get an error on the other side.

Reply:
ok looks like I can improve case 2 by extracting the error handler and setting it when before calling `this.send(res.stream)`

i.e.

```js
 const errorHandler = (err, request, reply) => {
      reply.request.log.warn(err, 'response errored')
        if (!reply.sent) {
          if (err.code === 'ERR_HTTP2_STREAM_CANCEL' || err.code === 'ENOTFOUND') {
            onError(reply, { error: new createError.ServiceUnavailable() })
          } else if (err instanceof TimeoutError || err.code === 'UND_ERR_HEADERS_TIMEOUT' || err.code === 'UND_ERR_BODY_TIMEOUT') {
            onError(reply, { error: new createError.GatewayTimeout() })
          } else {
            onError(reply, { error: createError(500, err) })
          }
        }
    }

    request({ method: req.method, url, qs, headers: requestHeaders, body }, (err, res) => {
      if (err) {
        errorHandler(err, this.request, this)
        return
      }
      this.request.log.info('response received')
      if (sourceHttp2) {
        copyHeaders(
          rewriteHeaders(stripHttp1ConnectionHeaders(res.headers)),
          this
        )
      } else {
        copyHeaders(rewriteHeaders(res.headers), this)
      }
      this.code(res.statusCode)
      if (onResponse) {
        onResponse(this.request.raw, this, res.stream)
      } else {
        this.context.errorHandler = errorHandler
        this.send(res.stream)
      }
    })
    return this
  })
```

Reply:
(note `errorHandler` doesn't fire on case 3 and I'm not sure there's much we could do other than terminating the response, I can add a test case for this)